src/Internal/BalanceTx/BalanceTx.purs

```diff
@@ -138,17 +138,11 @@ balanceTxWithConstraints
   -> QueryM (Either BalanceTxError FinalizedTransaction)
 balanceTxWithConstraints unbalancedTx constraintsBuilder =
   withBalanceTxConstraints constraintsBuilder $ runExceptT do
-    let
-      getWalletAddresses :: BalanceTxM (Array Address)
-      getWalletAddresses =
-        liftEitherQueryM $
-          QueryM.getWalletAddresses <#> note CouldNotGetWalletAddresses
-
     ownAddrs <-
-      maybe getWalletAddresses pure
+      maybe (liftQueryM QueryM.getWalletAddresses) pure
         =<< asksConstraints Constraints._ownAddresses
 
-    changeAddr <- liftMaybe CouldNotGetWalletAddresses $ Array.head ownAddrs
+    changeAddr <- liftMaybe CouldNotGetWalletAddress $ Array.head ownAddrs
```

Comment:
We need to think about which address we should select as `changeAddress` in case the user specifies the `mustBalanceTxWithAddress` constraint. My suggestion would be to make it explicit and provide an additional balancer constraint `mustSendChangeToAddress`. If this constraint is not set we would use the CIP-30 method. 
What do you think? @klntsky @jy14898 @Luis-omega

Reply:
Fully agree.

Reply:
https://github.com/Plutonomicon/cardano-transaction-lib/issues/1206